### ISSUE TITLE
SDU file format

### ISSUE THREAD

**Issue by Lantizia** [keirf/FlashFloppy#95]
Hey,

This is a feature request... I could try to supply a patch, but I'm not really familiar with this low level firmware stuff, and for someone familiar this should be an insanely simple change.

Basically I have an AST computer (remember those?) and they provided floppy disk image files on their recovery CDs... this was so you could recreate the floppies if you wished to install everything manually (e.g. DOS, Windows, MS Works, etc...)... rather than just wiping the whole machine and copying the filesystem straight from the CD with everything already installed.

These floppy image files end in extension .SDU and basically AST used the software from this guy to make them (so it's likely more than just AST users that have SDU files)...

http://www.saberman.com/SABDU/index.htm

On that site is a zip of his main program for making SDU files and includes some example SDU files.  It also has a .DLL for making your own end-user friendly extraction program (like on my recovery CD from my AST).  Here is the one from my CD (I've only included MS Video for Windows SDU files as at least that's freeware)... [astdu.zip](https://github.com/keirf/FlashFloppy/files/1887190/astdu.zip)

The short version of this story is... please support SDU files...

**All you have to do is ignore the first 46 bytes and then just treat them like a regular IMG file.**

p.s. I've just bought my drive off eBay (can't wait for it to arrive, it already has FlashFloppy on it - but I'll be sure to upgrade it, an OLED screen, buzzer and a rotary dial on it!)... so when it turns up and working I'll be sure to throw in a donation.

**Comment by keirf** [comment]
Not one I'm at all aware of, previously. If they are just IMG with a header then you can strip that yourself, on a machine with the `dd` command-line utility (eg Linux, MacOS):
```
dd if=my.dsu of=my.img bs=1 skip=46
```

**Comment by Lantizia** [comment]
Yeah I'm aware I can do that, I already do exactly that now.

But it'd be nice not to need to?  Admittedly it's a convenience thing, but it looks good to add another file format - if nothing else?

I am vaguely interested to see if anyone else finds this post who uses/used SABDU too.

**Comment by Lantizia** [comment]
Oh wow, quick turn around :)

Although my Gotek has technically arrived (with FlashFloppy already flashed)... it's this one with the cool black rotary selector (which I've not seen before - but is presumably supported in the default FlashFloppy firmware)... https://www.ebay.co.uk/itm/222905730623

I've gutted the components already so that they can go into another non-modified Gotek I've got coming in black (from Hong Kong, so a slow boat that's not arrived yet)...

Will get back to you.

**Comment by keirf** [comment]
Please reopen this ticket if issues arise in future testing.

**Comment by keirf** [comment]
I think SDU files must have zeros truncated from end of image file, which makes geometry detection (eg. double density vs high density) based on file size impossible.

I imagine the SDU header contains geometry info. Or else perhaps it only supports HD 1.44MB images? Do you have any other SDU images, ones which might not be HD 1.44MB?

**Comment by Lantizia** [comment]
Well here is all the ones I have (the sizes of the files do vary)
![image](https://user-images.githubusercontent.com/10448369/38855866-aa84b35c-421c-11e8-8533-dc616e85cc8e.png)

The first 46 bytes of them are all the same as each other (so the contents of those 46 bytes in all files will be the same as the 46 bytes you've got in the sample files I put in the astdu.zip above - as this full list of files below includes vfwr11d1.sdu and vfwr11d2.sdu).
![image](https://user-images.githubusercontent.com/10448369/38856347-15e35940-421e-11e8-912b-45c4f1e136bb.png)

I believe all the images were intended for HD 1.44MB floppy disks - I've never known it to be used for anything else.  The ASTDU program (their rebranded SABDU for customers) even shows this image which implies it too :)
![image](https://user-images.githubusercontent.com/10448369/38856483-8147360c-421e-11e8-9af4-6e513b8d72a9.png)

The SAD DU website (which based on its age - it's a miracle it is still up) doesn't seem to say one way or the other which kinds of disks it was meant to be used with... I'd probably need to use his software (the one that can author SDU files) to find it, will try that in a few days.

**Comment by Lantizia** [comment]
p.s. My last post has been heavily edited as it was wrong in maaany places, posting this to make sure if anyone has read the old one - re-read :D

**Comment by keirf** [comment]
I have emailed the author to ask about the SDU header format. For now, FlashFloppy v0.9.17a is hardcoded to 1.44MB PC geometry.

**Comment by Lantizia** [comment]
Brill - that comes up T:00.0 for both vfwr11d1.sdu and vfwr11d2.sdu now :)

Even mswareba.sdu and qkse3uk2.sdu work too, despite them having a different header for some reason (but presumably still the same length).

**Comment by keirf** [comment]
They may come up T:00.0 okay but not actually work. Looking at their header they have value 9 where most SDU images have 18. This would indicate a 9 sectors-per-track DD image rather than 18-spt HD. Actually the triplet (80, 2, 9) which could be (cyls, sides, spt) appears twice in the header -- I'm not sure why! I hope the SABDU author will help :)

**Comment by Lantizia** [comment]
I mailed him a few days back... just a one liner basically with a link to this page - in case he found it funny people were talking about this format/program again after many years.  This was before we got more in depth.

I did notice that sabdu@saberman.com bounces but sab@saberman.com didn't - so maybe there is someone still behind that address :D

**Comment by keirf** [comment]
I think for now, given that all images seen so far have the same (C,H,S) geometry triplet repeated in their header, I will just pick one arbitrarily and use that to set up the geometry in FlashFloppy.

**Comment by keirf** [comment]
Okay, please try this firmware, which auto-detects geometry based on SDU image header and probably works better for images mswareba.sdu and qkse3uk2.sdu:

[ff_95_2.zip](https://github.com/keirf/FlashFloppy/files/1919547/ff_95_2.zip)

**Comment by keirf** [comment]
I got a helpful email response from the SABDU author. I will revise the handling in FlashFloppy as necessary tomorrow and then close this ticket again after posting an update test firmware.

**Comment by Lantizia** [comment]
Wow that's great, any chance you can post what he said?

**Comment by keirf** [comment]
He tells me the last update of SAB DU was April 1st, 1995 :)

He also sent me some source code which has let me clean up the sdu handler a tiny bit (the code was already correct though, so not attaching another firmware here, just test the previous one, ff_95_2).
